enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+claim to be a "dual-stack" cloud, so it is assumed that both are supported.
+
+[6]:
+https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-multiple-ip-addresses-cli#add
+
+##### AWS
+
+The AWS API in [7] discusses additional private IP assignment to existing NICs.
+AWS seems to differentiate their API between [IPv4]:
+
+```
+aws ec2 assign-private-ip-addresses --network-interface-id $NIC_ID --private-ip-addresses 10.0.0.82
+```
+
+and [IPv6] assignments:
+
+```
+aws ec2 assign-ipv6-addresses --network-interface-id $NIC_ID --ipv6-addresses 2001:db8:1234:1a00:3304:8879:34cf:4071
+
+```
+
+[7]:
+https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/MultipleIP.html#ManageMultipleIP
+[IPv4]:
+https://docs.aws.amazon.com/cli/latest/reference/ec2/assign-private-ip-addresses.html
+[IPv6]:
+https://docs.aws.amazon.com/cli/latest/reference/ec2/assign-ipv6-addresses.html
+
+##### GCP
+
+GCP uses a different networking model, detailed in [8]. Instead of assigning
+additional IPs to NICs, they use something called "IP aliasing". Important to
+mention is that OpenShift cluster nodes on GCP usually get assigned a `/32`
+subnet on their primary interface. However, the IP aliasing subnet is `/19`
+(master nodes have: `10.0.0.0/19` and worker nodes: `10.32.0.0/19`). This
+facilitates this feature on GCP, as it means that we won't need to change the
+physical networking model on GCP as to make egress IP work. This however also
+means that we need to incorporate this fact into our design. The networking
+plugin - currently - only looks at the subnet mask on the primary interface
+defined on the host, when determining if an egress IP can be assigned to a given
+OpenShift node.
+
+The following shows the API on GCP for updating an instance and assigning it an
+additional IP address
+
+```
+gcloud compute instances network-interfaces update $INSTANCE_NAME \
+    --zone $ZONE \
+    [--network-interface $NIC; default="nic0"]
+    --aliases "$RANGE_NAME:$RANGE_CIDR,RANGE_NAME:RANGE_CIDR,..."
+```
+
+where
+
+- `$ZONE` the zone that contains the instance.
+- `$NIC` is the name of the network interface to which you're adding an alias IP
+  address range.
+- `$RANGE_NAME` the name of the subnet secondary range from which to draw the
+  alias IP range.
+- `$RANGE_CIDR` is the IP range to assign to the interface. The range can be a
+  specific range (192.168.100.0/24), a single IP address (192.168.100.1), or a
+  network mask in CIDR format (/24). If the IP range is specified by network
+  mask only, the IP allocator chooses an available range with the specified
+  network mast and allocates it to the network interface.
+
+Note: an egress IP setup on GCP has already been accomplished using
+OVN-Kubernetes in a proof of concept / experimental mode. A couple of things
+were noticed during that experiment:
+
+- the alias IP assigned does not have any impact on the host, meaning: once the
```

Comment:
How do other cloud providers handle egress IP assignment to the nic? In other words, do we need to do some equivalent of `ip addr add` on them as well for assigning the egress IP to the nic?

Reply:
Not sure yet as I just did that POC on GCP for the moment. We have https://issues.redhat.com/browse/SDN-1431 and similar stories to investigate this before starting the implementation.

Reply:
This is the expected behavior on all clouds. The cloud API has no ability to affect OS-level aspects of the VMs. When you allocate an IP to a node, you are only updating the routing/filtering rules on the network between the VMs, not changing anything about the VMs themselves.